src/events/balancer.py

```diff
@@ -0,0 +1,117 @@
+"""Balancer role core charm logic."""
+
+import logging
+from subprocess import CalledProcessError
+from typing import TYPE_CHECKING
+
+from ops import (
+    ActiveStatus,
+    ConfigChangedEvent,
+    EventBase,
+    Object,
+    UpdateStatusEvent,
+)
+
+from literals import (
+    BALANCER,
+    Status,
+)
+from managers.balancer import BalancerManager
+from managers.config import BalancerConfigManager
+from workload import BalancerWorkload
+
+if TYPE_CHECKING:
+    from charm import KafkaCharm
+
+logger = logging.getLogger(__name__)
+
+
+class BalancerOperator(Object):
+    """Implements the logic for the balancer."""
+
+    def __init__(self, charm) -> None:
+        super().__init__(charm, BALANCER.value)
+        self.charm: "KafkaCharm" = charm
+
+        self.workload = BalancerWorkload()
+
+        # Fast exit after workload instantiation, but before any event observer
+        if BALANCER.value not in self.charm.config.roles:
+            return
```

Comment:
**question:** I think this is a fairly sane way to do it. What about moving it to `charm.py` so it doesn't even need to reach this point?

Reply:
I wanted `charm.py` to be completely unaware of what sub operators are running so that we don't have to use defensive programming everywhere.
Please let me know if I answered your question correctly